### DISCUSSION TITLE
Questions about the neuron code

### DISCUSSION THREAD

**Discussion by Xxin980219** [jeshraghian/snntorch#212] [Q&A]
* snntorch version:
* Python version:
* Operating System:

### Description
SNN neurons, e.g., the code of the Leaky neuron model, where the forward propagation process that determines whether a neuron is fired by a pulse should be the one in which the membrane potential at the previous moment and the input at this moment jointly influence whether a pulse is fired, rather than the membrane potential at the previous moment alone determining whether a pulse is fired. In addition, some of the code of the state update of Alpha, Leaky, RLeaky,Synaptic neuron models do not conform to the update formula of the document, such as the code of the subtraction mechanism in Leaky's (_build_state_function） function, which I guess may be an error in the position of the brackets.

**Comment by jeshraghian** [comment]
This was recently updated though there is a chance some errors weren't caught. 
Are you able to specify the differences?

**Comment by Xxin980219** [comment]
I apologize for asking a hasty question without understanding the code before. Next I will talk about my understanding, can you help me to see if it is correct?
![image](https://user-images.githubusercontent.com/94166823/233760281-b42ecf66-4714-4172-b97c-8679e4fa476a.png)
The figure is from the article **"Training Spiking Neural Networks Using Lessons From Deep Learning"**. We can see that the membrane potential U[t] is determined by the inputs X[t] ,U[t-1] and Sout[t-1],the calculation equation is shown below.
![image](https://user-images.githubusercontent.com/94166823/233770537-15abe487-6034-4934-83aa-b8db3e87ad1b.png)

### The following is part of the code for forward propagation of neurons:

      def forward(self, input_, mem=False):
           ......
           if not self.init_hidden:
            self.reset = self.mem_reset(mem)
            mem = self._build_state_function(input_, mem)

            if self.state_quant:
                mem = self.state_quant(mem)

            if self.inhibition:
                spk = self.fire_inhibition(mem.size(0), mem)  # batch_size
            else:
                spk = self.fire(mem)

            return spk, mem
           ......

         def _build_state_function(self, input_, mem):
                if self.reset_mechanism_val == 0:  # reset by subtraction
                    state_fn = self._base_state_function(
                        input_, mem - self.reset * self.threshold
                    )
                elif self.reset_mechanism_val == 1:  # reset to zero
                    state_fn = self._base_state_function(
                        input_, mem
                    ) - self.reset * self._base_state_function(input_, mem)
                elif self.reset_mechanism_val == 2:  # no reset, pure integration
                    state_fn = self._base_state_function(input_, mem)
                return state_fn
From this code, it is divided into three steps:
1.  the neuron receives the unreset  membrane potential of the previous moment and  determines whether it reaches the threshold of being able to fire the spiking.
2.  then update the membrane potential at this moment. 
3.  finally, the neuron determines whether or not it is able to fire a spiking at this moment and returns the spiking and the unreset membrane potential at this moment.

**So the final output is the spikings and the unreset membrane potential generated by all neurons.**

**Comment by jeshraghian** [comment]
Converting this to a discussion.

1. `self.reset = self.mem_reset(mem)` checks if the membrane is above the threshold. If so, then this neuron should have spiked at the previous time step. This time, the threshold check is used to trigger `reset=1`.

2. Then update membrane potential. This uses the new reset term.

3. Then as you have said, the new membrane potential is used to determine if there is another spike. 

Hope this makes sense.
